homeassistant/components/switchbot/config_flow.py

```diff
@@ -0,0 +1,180 @@
+"""Config flow for Switchbot."""
+from __future__ import annotations
+
+from asyncio import Lock
+import logging
+from typing import Any
+
+from switchbot import GetSwitchbotDevices  # pylint: disable=import-error
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry, ConfigFlow, OptionsFlow
+from homeassistant.const import CONF_MAC, CONF_NAME, CONF_PASSWORD, CONF_SENSOR_TYPE
+from homeassistant.core import callback
+from homeassistant.data_entry_flow import FlowResult
+
+from .const import (
+    ATTR_BOT,
+    CONF_RETRY_COUNT,
+    CONF_RETRY_TIMEOUT,
+    CONF_SCAN_TIMEOUT,
+    CONF_TIME_BETWEEN_UPDATE_COMMAND,
+    DEFAULT_RETRY_COUNT,
+    DEFAULT_RETRY_TIMEOUT,
+    DEFAULT_SCAN_TIMEOUT,
+    DEFAULT_TIME_BETWEEN_UPDATE_COMMAND,
+    DOMAIN,
+)
+
+_LOGGER = logging.getLogger(__name__)
+CONNECT_LOCK = Lock()
+
+
+def _btle_connect(mac: str) -> dict:
+    """Scan for BTLE advertisement data."""
+    # Try to find switchbot mac in nearby devices,
+    # by scanning for btle devices.
+
+    switchbots = GetSwitchbotDevices()
+    switchbots.discover()
+    switchbot_device = switchbots.get_device_data(mac=mac)
+
+    if not switchbot_device:
+        raise NotConnectedError("Failed to discover switchbot")
+
+    return switchbot_device
+
+
+class SwitchbotConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Switchbot."""
+
+    VERSION = 1
+
+    async def _validate_mac(self, data: dict) -> FlowResult:
+        """Try to connect to Switchbot device and create entry if successful."""
+        await self.async_set_unique_id(data[CONF_MAC].replace(":", ""))
+        self._abort_if_unique_id_configured()
+
+        # Validate bluetooth device mac.
+        # CONNECT_LOCK prevents btle adapter exceptions if there are multiple calls to this method.
+        try:
+            async with CONNECT_LOCK:
```

Comment:
Store this in `hass.data[DOMAIN][BTLE_LOCK]` instead since it likely needs to share the same lock with the coordinator.

Reply:
Done.